### PR TITLE
Add APIv4 for RecentItems

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Exposes the list of recently viewed items to the api.

Before
----------------------------------------
Extensions like https://github.com/civicrm/org.civicrm.recentmenu must directly use Civi functions.

After
----------------------------------------
They can use the API.

Technical Details
----------------------------------------
This also makes it easier to add a recent item to the list, requiring fewer calculations as many defaults (title, icon, and links) are now set automatically.

### CODE REVIEW

File: CRM/Contact/BAO/Contact.php [hunk 1]
```diff
@@ -752,7 +752,7 @@ protected static function contactTrash($contact): bool {
     $contact->save();
     CRM_Core_BAO_Log::register($contact->id, 'civicrm_contact', $contact->id);
 
-    CRM_Utils_Hook::post('update', $contact->contact_type, $contact->id, $contact);
+    CRM_Utils_Hook::post('edit', $contact->contact_type, $contact->id, $contact);
```
Comment (colemanw) [inline CRM/Contact/BAO/Contact.php:755]: @eileenmcnaughton git blames you but you didn't write this you just moved it. But FYI it was wrong - there is no such action as "update" for pre/post hooks.

Comment (eileenmcnaughton) [reply CRM/Contact/BAO/Contact.php:755]: I blame git right back!

**Comment by eileenmcnaughton** [comment]
@colemanw the top four are with this patch & the bottom 2 are before this patch

image

This is a newly rebuild site and I note that the contact-type table looks like this - 

image

My 5.48 site looks like

image

**Comment by colemanw** [comment]
@eileenmcnaughton the problem was that `regen.sh` hadn't been run since #23073 so the contact type icons were missing.
I've run it and pushed it to this PR (retest this please).

**Comment by colemanw** [comment]
`test-3` seems to be out of disk space. Retest this please

**Comment by colemanw** [comment]
Retest this please

**Comment by colemanw** [comment]
Retest this please

**Comment by colemanw** [comment]
Retest this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
Ok this is now looking like

image


I think it is mergeable

File: CRM/Utils/Recent.php [hunk 2]
```diff
@@ -168,11 +233,23 @@ private static function getIcon($type, $subType) {
...
+      if ($event->action === 'delete') {
+        // Is this an entity that might be in the recent items list?
+        $providersPermitted = Civi::settings()->get('recentItemsProviders') ?: array_keys(self::getProviders());
+        if (in_array($entityType, $providersPermitted)) {
+          self::del(['entity_id' => $event->id, 'entity_type' => $entityType]);
+        }
+      }
+      elseif ($event->action === 'edit') {
+        if (isset($event->object->is_deleted)) {
+          \Civi\Api4\RecentItem::update()
```
Comment (mattwire) [inline CRM/Utils/Recent.php:247]: @colemanw This seems to have caused a regression with webform_civicrm submitting activities/cases via user checksum. Adding FALSE in checkPermissions here fixes that.

Comment (eileenmcnaughton) [reply CRM/Utils/Recent.php:247]: - I suspect the permissions for that api might be too tight
-

Comment (colemanw) [reply CRM/Utils/Recent.php:247]: I think the permissions are ok as "access CiviCRM', but yea for a checksum user who has no permissions at all... womp womp.

**Pull request merged**